### ISSUE TITLE
[OJS] Slow query in issueDao getNumArticles

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#6777]
Affecting at least 3.2.1 stable, probably also 3.3 with a large database.

When the Back issues listing in the dashboard is loaded, the `getNumArticles` is called for each listed issue.
The query is very slow in larger databases taking between 0.7 to 1.5 seconds to run for each individual issue, resulting into a very slow back issues listing. In a large journal the listing can load up to 2-3 minutes.

The reason is that the sql is missing apostrophes meaning that the `issueId` value passed there is probably considered as an integer but the field is a text field. Mysql probably has to do casting before it can use the given value and it can not use the index containing the setting_name and setting_value.

Changing the query to this will make the single sql to run in 0.02 seconds and the same Back issues page opens in 1-3 seconds:

```
		$result = $this->retrieve(
			"SELECT COUNT(*)
				FROM submissions s
				LEFT JOIN publications p ON (p.publication_id = s.current_publication_id)
				LEFT JOIN publication_settings ps ON (ps.publication_id = p.publication_id)
				WHERE ps.setting_name = ? AND ps.setting_value = '?' AND (s.status = ? OR s.status = ?) ",
			['issueId', (int) $issueId, (int) STATUS_SCHEDULED, (int) STATUS_PUBLISHED]
		);
```
If you know a better way of passing the apostrophes to the query let me know.

ping @jonasraoni and @henriqueramos as we have already discussed this in Slack

also ping @NateWr and @asmecher, I think we have to track down other sql queries that might have the some problem? For example thinking if this could had been fixed just by adding apostrophes https://github.com/pkp/ojs/commit/4e6c58825ce236dd23b8e99155a6f204dc17a3b5#diff-ab9b11f252d00204109cbd06dffcd8c946a2ef6e094bf3a6402c61052dbd47d3

**Comment by ajnyga** [comment]
Actually this is actually the better way of doing it:

```
		$result = $this->retrieve(
			'SELECT COUNT(*)
				FROM submissions s
				LEFT JOIN publications p ON (p.publication_id = s.current_publication_id)
				LEFT JOIN publication_settings ps ON (ps.publication_id = p.publication_id)
				WHERE ps.setting_name = ? AND ps.setting_value = \'?\' AND (s.status = ? OR s.status = ?) ',
			['issueId', (int) $issueId, (int) STATUS_SCHEDULED, (int) STATUS_PUBLISHED]
		);

```

**Comment by asmecher** [comment]
@ajnyga, see the recent(ish) work on https://github.com/pkp/pkp-lib/issues/6320 -- I don't think variable binding will work in quotes as you've got itas you've got it; part of the problem is also the need for the cast (however it's done) to work in both MySQL and PostgreSQL.

But fundamentally I think the use of `setting_value` to store a numeric ID is always going to be a problem. Perhaps we should consider re-adding an `issue_id` column to `publications`, or adding a `setting_value_numeric` column to `publication_settings` for cases when we want something numeric?

**Comment by NateWr** [comment]
A third approach would be to make a move towards the [submission groups proposal](https://github.com/pkp/pkp-lib/issues/4415) and create a separate table where the issue / submission relationships are stored. (This would entail moving the issue assignment from publications to submissions -- I may have been overeager to move some data to publications that would be better managed at the submission level.)

**Comment by ajnyga** [comment]
I am not sure what the best solution is but this basically prevents larger installations from using OJS 3.2+ so it should be addressed fairly quickly.

I think moving the submissionGroup/issue relation to submission could be problematic. I can see situations where a new version of a article is moved to another category for example.

**Comment by henriqueramos** [comment]
> @ajnyga, see the recent(ish) work on #6320 -- I don't think variable binding will work in quotes as you've got itas you've got it; part of the problem is also the need for the cast (however it's done) to work in both MySQL and PostgreSQL.
> 
> But fundamentally I think the use of `setting_value` to store a numeric ID is always going to be a problem. The database is always going to have to cast during join which is going to be inefficient/cranky. Perhaps we should consider re-adding an `issue_id` column to `publications`, or adding a `setting_value_numeric` column to `publication_settings` for cases when we want something numeric?

Anyone tried to use CAST() function ([mysql](https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html#function_cast) and [postgresql](https://www.postgresql.org/docs/9.2/sql-createcast.html) docs)?

**Comment by ajnyga** [comment]
> For comparisons of a string column with a number, MySQL cannot use an index on the column to look up the value quickly. If str_col is an indexed string column, the index cannot be used when performing the lookup in the following statement:
> 
> SELECT * FROM tbl_name WHERE str_col=1;
> 
> The reason for this is that there are many different strings that may convert to the value 1, such as '1', ' 1', or '1a'.

https://dev.mysql.com/doc/refman/8.0/en/type-conversion.html

**Comment by jonasraoni** [comment]
I'm not following this discussion deeply, so I'll just leave some generic comments =]

1. If I remember well, @ajnyga said the query was being called **inside a loop**, so I think it deserves a _hot path_, in the outer scope. Which could be some kind of cache or to join the queries into a single statement (it's faster to transfer some duplicated data than doing extra database work).

2. About the `*_settings` table:
- I think `system attributes` should be kept in fixed columns, with the right data type, which will also help with the performance. So, why the `issueId` is living in the publication_settings?

- AFAIK these tables are used mostly for localization, so I think we could use another structure for that. Taking the `user_settings` as an example, we could have errrrr: `user_localized (locale varchar(14), givenName varchar(255), familyName varchar(255), ...)`.
I think it's ok to keep the user_settings alive, but it should be used for very specific cases, such as untyped information (e.g. external data coming from a JSON) or to be used by plugins.

**Comment by ajnyga** [comment]
Noticed today that the fix I propose above actually broke the count (although was fast)

This is how I fixed it now. It is now both fairly fast and also actually works. I basically cast the issueId as (string). This works in mysql, not sure what happens in postgre.

```
		$result = $this->retrieve(
			'SELECT COUNT(*)
				FROM submissions s
				LEFT JOIN publications p ON (p.publication_id = s.current_publication_id)
				LEFT JOIN publication_settings ps ON (ps.publication_id = p.publication_id)
				WHERE ps.setting_name = ? AND ps.setting_value = ? AND (s.status = ? OR s.status = ?) ',
			['issueId', (string) $issueId, (int) STATUS_SCHEDULED, (int) STATUS_PUBLISHED]
		);
```

(Funny thing with this one, was confused why my issueDao is different from the stable branch, had totally forgot I did this change to my upgrade test OJS 🤪

**Comment by asmecher** [comment]
That'll work for PostgreSQL without problems, at a glance. Mind opening a PR for it?

**Comment by asmecher** [comment]
@ajnyga, this was solved separately by just casting an int to a string in PHP-land. I'm a little surprised that the type of binding in the host language would affect the optimizer in the DBMS -- some reading for me to do. It looks like this resolves it:

https://github.com/pkp/pkp-lib/pull/7483

...but confirmation from your side would be welcome!

**Comment by NateWr** [comment]
@asmecher I noticed that this fix is applied to `main` but modifies a 3.3.0 migration step. Do we need to do some additional work to fix this for OJS instances that have already upgraded to 3.3.0-x?

**Comment by asmecher** [comment]
Thanks, I read that too fast, @NateWr -- this is the same problem but applies to a different part of the code. Applied to the `getNumArticles` function referred to in this entry, it's: https://github.com/pkp/ojs/commit/7c885b1fdcb7f22620ac30c2ec56474581e92230
